game/src/main/kotlin/world/gregs/voidps/world/activity/dnd/shootingstar/ShootingStar.kts

```diff
@@ -0,0 +1,332 @@
+    )
+}
+
+objectOperate("Mine") {
+    val isStar: Boolean = target.def["star", false]
+    if(!isStar){
+        return@objectOperate
+    }
+    if (target.id.startsWith("depleted")) {
+        player.message("There is currently no ore available in this rock.")
+        return@objectOperate
+    }
+    player.softTimers.start("mining")
+    var first = true
+    while (true) {
+        if (!objects.contains(target)) {
+            break
+        }
+
+        if (player.inventory.isFull()) {
+            player.message("Your inventory is too full to hold any more ore.")
+            break
+        }
+
+        val rock: Rock? = target.def.getOrNull("mining")
+        if (rock == null || !player.has(Skill.Mining, rock.level, true)) {
+            break
+        }
+        val pickaxe = getBestPickaxe(player)
+        if (!hasRequirements(player, pickaxe, true) || pickaxe == null) {
+            break
+        }
+        val delay = if (pickaxe.id == "dragon_pickaxe" && random.nextInt(6) == 0) 2 else pickaxe.def["mining_delay", 8]
+        if (first) {
+            player.message("You swing your pickaxe at the rock.", ChatType.Filter)
+            first = false
+        }
+        val remaining = player.remaining("skill_delay")
+        if (remaining < 0) {
+            player.face(target)
+            player.setAnimation("${pickaxe.id}_swing_low")
+            player.start("skill_delay", delay)
+            pause(delay)
+        } else if (remaining > 0) {
+            pause(delay)
+        }
+        if (rock.gems) {
+            val glory = player.equipped(EquipSlot.Amulet).id.startsWith("amulet_of_glory_")
+            if (Level.success(player.levels.get(Skill.Mining), if (glory) 3..3 else 1..1)) {
+                addStarDust(player, gems.random())
+                continue
+            }
+        }
+        var ores = rock.ores
+        if (target.id == "rune_essence_rocks") {
+            val name = if (World.members && player.has(Skill.Mining, 30)) "pure_essence" else "rune_essence"
+            ores = rock.ores.filter { it == name }
+        }
+        for (item in ores) {
+            val ore = itemDefinitions.get(item)["mining", Ore.EMPTY]
+            if (Level.success(player.levels.get(Skill.Mining), ore.chance)) {
+                player.experience.add(Skill.Mining, ore.xp)
+                totalCollected++
+                handleMinedStarDust(target)
+                val totalStarDust = player.inventory.count(item) + player.bank.count(item)
+                if(totalStarDust >= ownedStarDustThreshold){
+                    player.message("You have the maximum amount of star dust but was still rewarded experience.")
+                } else {
+                    if (!addStarDust(player, item)) {
+                        player.clearAnimation()
+                        break
+                    }
+                }
+            }
+        }
+        player.stop("skill_delay")
+    }
+    player.softTimers.stop("mining")
+}
```

Comment:
Okay so I think the actual mining bit can be done inside `Mining.kts` rather than copy it all.

```kotlin
fun addOre(player: Player, ore: String): Boolean {
    if (ore == "stardust") {
        val totalStarDust = player.inventory.count(item) + player.bank.count(item)
        if (totalStarDust >= ownedStarDustThreshold){
            player.message("You have the maximum amount of star dust but was still rewarded experience.")
            return true
        }
    }
    val added = player.inventory.add(ore)
    if (added) {
        player.message("You manage to mine some ${ore.toLowerSpaceCase()}.")
    } else {
        player.inventoryFull()
    }
    return added
}
```

```kotlin
fun deplete(rock: Rock, obj: GameObject): Boolean {
    if (obj.id.startsWith("crashed_star_tier_")) {
        val finished = handleMinedStarDust(obj)
        return finished
    }
    if (rock.life >= 0) {
        objects.replace(obj, "depleted${obj.id.dropWhile { it != '_' }}", ticks = rock.life)
        return true
    }
    return false
}
```

Reply:
The reason why i had copied the whole mining action from 'Mining.kts' to 'ShootingStar.kts' was because of this function:

```kotlin
fun handleMinedStarDust(currentMinedStar: GameObject) {
    val starPayout = currentMinedStar.def["collect_for_next_layer", -1]
    if(totalCollected >= starPayout){
        val stage = currentMinedStar.id.takeLast(1)
        if(stage.equals("1")){
            currentMinedStar.remove()
            cleanseEvent(false)
            return
        }
        val nextStage = currentMinedStar.id.replace(stage, (stage.toInt() - 1).toString())
        val nextstar = currentMinedStar.replace(nextStage)
        totalCollected = 0
        changeStar(currentMinedStar.id, nextstar.id)
    }
}

```
I am still pretty new to kotlin scripting and wasn't sure how i would run a func in one script from another script.